### ISSUE TITLE
Doesn't run on colab with Pygmalion-6B / results look different on Colab

### ISSUE THREAD

**Issue by waifusd** [oobabooga/text-generation-webui#14]
Using the provided notebook and just changing the model to Pygmalion-6B instead of Pygmalion-1.3B generates the following tcmalloc error and the execution stops cold.

![tcmalloc](https://user-images.githubusercontent.com/123158074/213659678-ecdfaf31-697c-4c1b-9c7c-90c9ff12689a.png)

**Comment by BushyToaster88** [comment]
> Using the provided notebook and just changing the model to Pygmalion-6B instead of Pygmalion-1.3B generates the following tcmalloc error and the execution stops cold.
> 
> ![tcmalloc](https://user-images.githubusercontent.com/123158074/213659678-ecdfaf31-697c-4c1b-9c7c-90c9ff12689a.png)

I have the same issue when trying to run the 4chan model in colab. Im get a "Memory cgroup out of memory" error when I look at the dmesg in the colab terminal.

**Comment by oobabooga** [comment]
I can confirm this issue. 

On colab, I can't load either pygmalion-2.7b or pygmalion-6b.

The free colab has around 13GB RAM, while pygmalion-2.7b takes 6.8GB of RAM to load on my system (peak allocation). So it should in principle work.

**Comment by waifusd** [comment]
Pygmalion-6B should be able to loaded on colab too, since the colab notebooks of other projects are able to load it (namely the Pyg devs own notebook, and KoboldAI's).

**Comment by oobabooga** [comment]
An Anonymous 4chan user has kindly provided this notebook that allows the 6b model to be loaded in 8bit mode:

https://colab.research.google.com/github/81300/AI-Notebooks/blob/main/Colab-TextGen-GPU.ipynb

I haven't tested it yet because Google is not giving me a free instance with a GPU.

**Comment by waifusd** [comment]
I tried it:

![pygoob1](https://user-images.githubusercontent.com/123158074/213779900-af9f4ab8-ef97-4f2a-813e-62805cc9b607.png)

**Comment by oobabooga** [comment]
I can confirm that the results look worse on colab than running locally. I made a comparison using `debug.txt`, which is a deterministic preset that should generate the same responses:

```
do_sample=False,
max_new_tokens=tokens,

```

But the responses are different:

| Colab | Local |
| ------------- | ------------- |
| ![colab2](https://user-images.githubusercontent.com/112222186/213812001-4357ccda-4045-41d7-9382-9a42a523c58e.png)  | ![local2](https://user-images.githubusercontent.com/112222186/213812008-b08a818f-6056-4a99-869a-5787b64c4455.png)  |

I don't know the cause and will leave this issue open to see if someone has an idea.

Some other observations:

* It is now possible to load the 6b model with `!python server.py --cai-chat --share --load-in-8bit` or `!python server.py --cai-chat --share --auto-devices` (after the bitsandbytes version upgrade suggested by anon). The tcmalloc warnings still appear, but the model loads successfully. 
* My instance had a Tesla T4 GPU.

**Comment by waifusd** [comment]
I still can't launch the "basic commands" colab even with the additional arguments.

I tried a little more on the Anon's colab with a different character and without fail, within the first 5 messages, I can always get the bot to sperg out with a normal chat:

![image](https://user-images.githubusercontent.com/123158074/213861743-8fb1942a-bcd4-45d2-8048-06a7570a0d8e.png)

Wish I could test a local installation but without a GPU I can only do colab.

**Comment by oobabooga** [comment]
@81300 thanks for looking into this. The most likely culprit is indeed the CUDA library. The debug preset generating different results implies that the logits are different, which can only happen if the internal calculations are performed differently (different precision?).

`--no-stream` shouldn't change the results at all. It should only make the speed of text generation a bit slower. Try uploading a file called `debug.txt` under `text-generation-webui/presets` with the settings on my previous response and you will be able to verify that the results are the same when you run with or without streaming.

In any case, I don't know why, but the replies in your notebook are a LOT better now. I have sent 18 messages to Chiharu with the default settings and she didn't start ranting once.

![out](https://user-images.githubusercontent.com/112222186/213890164-3f3424f3-4f6d-421d-aa66-b0913e7f45fe.png)

**Comment by oobabooga** [comment]
My only remaining question is if it is possible to get the exact same responses on colab and locally. The colab responses feel a bit worse than local. Not nearly as bad as before, but still not as good.

The debug preset is now included by default.

**Comment by waifusd** [comment]
Yesterday I tried the colab again with Pyg-6B, the bot didn't start to rant in the first 5 messages, and the answers were "slightly passable" at first (I've never been able to run it locally so I have no idea how that's supposed to look, so my only point of comparison is CAI, to which Pyg is not even remotely close to).
After 20 or so messages the bot started to answer with messages that were 80% identical to the last one, and it culminated to it responding with a 1:1 identical message. Tried regenerating or generating a message while staying silent (not inputting a message of my own) and the bot would spout the same identical message.

**Comment by oobabooga** [comment]
It is also possible that the divergent results are caused by `device_map='auto'` and not the GPU architecture, see this issue:

https://github.com/huggingface/transformers/issues/20896

**Comment by oobabooga** [comment]
Here is the 

>Hi

test on my RTX 3090 with the debug preset, for comparison. On my laptop which has a turing gpu the results are the same as Colab and @81300.

The goal is to reproduce this >Hi result on any GPU.

![print](https://user-images.githubusercontent.com/112222186/214425632-22412814-8baf-4e5f-88d6-221e865d7c32.png)

**Comment by oobabooga** [comment]
@81300 @waifusd, I have discovered the issue: **it was the model**.

The model that I was using on my computer was the very first commit to the [HuggingFace repository](https://huggingface.co/PygmalionAI/pygmalion-6b), which I downloaded on January 12nd.

The current commit to that repository (main branch) is different from the first one. This updated commit passes the >Hi test when executed in GPU-only mode, but fails in CPU, GPU+CPU, or 8-bit mode, generating the "what do you think of my setup?" response that we have been seeing.

On the other hand, the first commit passes the >Hi test in any mode. It always yields the same responses. This is the response that I got on colab using this commit:

![colab](https://user-images.githubusercontent.com/112222186/214462868-587a464a-30aa-4075-8303-39ef84957f24.png)

I have re-uploaded this first commit here: https://huggingface.co/oobabooga/pygmalion-6b-original

You can try it in this notebook (which is the one I used for the screenshot above): https://colab.research.google.com/github/oobabooga/AI-Notebooks/blob/main/Colab-TextGen-GPU.ipynb

In other words, it is now possible to get 1000x better responses on Colab.

**Comment by oobabooga** [comment]
@81300: you are right that the next step now would be to ditch 8-bit mode altogether, as this would probably make the model run a bit faster.

I am also worried about the loading times on Colab. Ideally, it would be best to get the model working in less than 5 minutes instead of 12.

**Comment by oobabooga** [comment]
Here are some comparisons between different branches of pygmalion-6b and different modes (GPU, GPU+CPU, and 8-bit):

https://huggingface.co/PygmalionAI/pygmalion-6b/discussions/8#63d15cae119416cdbe15ae2e
